torch/csrc/autograd/profiler_python.cpp

```diff
@@ -0,0 +1,707 @@
+// `std::variant`? (Or `c10::variant`, as it were.) The answer is that due
+// to alignment the tag would have to be packed with the union data and
+// `RawEvent` would grow to three words. (Not just 50% bigger, but also less
+// cache friendly.)
+struct RawEvent {
+    RawEvent(TraceTag tag, int lasti, TraceContext* ctx)
+            : tag_(static_cast<uint8_t>(tag)),
+              thread_id_(ctx->thread_id_),
+              lasti_(static_cast<uint16_t>(lasti)),
+              misc_() {
+        int64_t t = now() - ctx->initial_us_;
+        t_ = static_cast<uint32_t>(t);
+
+        TORCH_INTERNAL_ASSERT_DEBUG_ONLY(lasti <= std::numeric_limits<uint16_t>::max());
+        TORCH_INTERNAL_ASSERT_DEBUG_ONLY(t <= std::numeric_limits<uint32_t>::max());
+    }
+
+    RawEvent(TraceTag tag, int lasti, TraceContext* ctx, PyCodeObject* f_code)
+            : RawEvent(tag, lasti, ctx) {
+        TORCH_INTERNAL_ASSERT_DEBUG_ONLY(tag == TraceTag::kPy_Call);
+        misc_.f_code_ = f_code;
+    }
+
+    RawEvent(TraceTag tag, int lasti, TraceContext* ctx, PyObject* arg)
+            : RawEvent(tag, lasti, ctx) {
+        TORCH_INTERNAL_ASSERT_DEBUG_ONLY(tag == TraceTag::kC_Call);
+        misc_.arg_ = arg;
+    }
+
+    uint8_t tag_;
+    uint8_t thread_id_;
+    uint16_t lasti_;
+    uint32_t t_;
+    union {
+        // TraceTag::kPy_Call
+        PyCodeObject* f_code_;
+
+        // TraceTag::kC_Call
+        PyObject* arg_;
+
+        // TraceTag::kPy_Return
+        // TraceTag::kC_Return
+        // ** Unused (placeholder) **
+        void* null_;
+    } misc_;
+
+    TraceTag tag() const {
+        return static_cast<TraceTag>(tag_);
+    }
+
+    int lasti() const {
+        // f_lasti is positive, with one exception: CPython intializes frames
+        // with `f_lasti = -1`. We don't want to give up half of the range by
+        // switching to int16_t. So instead we do the fast (underflowing) cast
+        // in the ctor, and rectify the value in this accessor which should
+        // only be called during trace post processing.
+        return lasti_ == std::numeric_limits<uint16_t>::max()
+            ? (int)(-1)
+            : (int)lasti_;
+    }
+};
+
+static_assert(sizeof(RawEvent) == 16, "RawEvent should be exactly 16 bytes.");
+
+
+// std::hash doesn't have a specialization for pairs so we have to define one.
+// A simple XOR is good enough for our purposes.
+struct hash_pair {
+    template <class T1, class T2>
+    size_t operator() (const std::pair<T1, T2>& pair) const {
+        return std::hash<T1>()(pair.first) ^ std::hash<T2>()(pair.second);
+    }
+};
+
+
+// ============================================================================
+// == Tracing implementation ==================================================
+// ============================================================================
+constexpr size_t max_py_threads = std::numeric_limits<uint8_t>::max() + 1;
```

Comment:
is it possible to even hit this many threads? why not just do something simple like `64`?

Reply:
For something I/O heavy maybe. I figure I may as well go up to what the bits can represent.